servers/rendering/renderer_rd/scene/texture_rd.cpp

```diff
@@ -0,0 +1,340 @@
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "texture_rd.h"
+
+////////////////////////////////////////////////////////////////////////////
+// Texture2DRD
+
+void Texture2DRD::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_texture_rd_rid", "texture_rd_rid"), &Texture2DRD::set_texture_rd_rid);
+	ClassDB::bind_method(D_METHOD("get_texture_rd_rid"), &Texture2DRD::get_texture_rd_rid);
+
+	ADD_PROPERTY(PropertyInfo(Variant::RID, "texture_rd_rid"), "set_texture_rd_rid", "get_texture_rd_rid");
+}
+
+int Texture2DRD::get_width() const {
+	return size.width;
+}
+
+int Texture2DRD::get_height() const {
+	return size.height;
+}
+
+RID Texture2DRD::get_rid() const {
+	if (texture_rid.is_null()) {
+		//we are in trouble, create something temporary
+		texture_rid = RenderingServer::get_singleton()->texture_2d_placeholder_create();
+	}
+
+	return texture_rid;
+}
+
+bool Texture2DRD::has_alpha() const {
+	return false;
+}
+
+Ref<Image> Texture2DRD::get_image() const {
+	ERR_FAIL_NULL_V(RS::get_singleton(), Ref<Image>());
+	if (texture_rid.is_valid()) {
+		return RS::get_singleton()->texture_2d_get(texture_rid);
+	} else {
+		return Ref<Image>();
+	}
+}
+
+void Texture2DRD::set_texture_rd_rid(RID p_texture_rd_rid) {
+	ERR_FAIL_NULL(RS::get_singleton());
+
+	if (p_texture_rd_rid.is_valid()) {
+		ERR_FAIL_NULL(RD::get_singleton());
+		ERR_FAIL_COND(!RD::get_singleton()->texture_is_valid(p_texture_rd_rid));
+
+		RD::TextureFormat tf = RD::get_singleton()->texture_get_format(p_texture_rd_rid);
+		ERR_FAIL_COND(tf.depth > 1);
+		ERR_FAIL_COND(tf.array_layers > 1);
+
+		size.width = tf.width;
+		size.height = tf.height;
+		// TODO add check to see which formats have alpha... Is this even important?
```

Comment:
Looks like most texture types just return false. This was likely used a long time ago, but isn't functional anymore. I would leave it as is (with comment removed) and then we can add a check if needed